### PR TITLE
nixosTest: document wait for stdout behavior, fix #144613

### LINKED ISSUE #144613: nixosTests.pt2-clone and others hang
### Describe the bug

`pt2-clone.tests`/`nixosTests.pt2-clone`, `nixosTests.ft2-clone`, `nixosTests.shattered-pixel-dungeon`, `nixosTests.xterm` (and probably others) hangs:

```
[...]
machine # [   25.071572] systemd[952]: Reached target Current graphical user session.
(27.39 seconds)
machine # [   26.830709] systemd-udevd[687]: controlC0: Process '/nix/store/i1jkln6xh32aw11xmjk4p8f14cpzszjz-alsa-utils-1.2.5.1/sbin/alsactl restore 0' failed with exit code 99.
machine # [   26.889435] systemd[1]: Reached target Sound Card.
machine # [   26.918075] systemd[952]: Reached target Sound Card.
^Cerror: interrupted by the user
```

Full log: [pt2-clone.tests.txt](https://github.com/NixOS/nixpkgs/files/7474917/pt2-clone.tests.txt)

### Steps To Reproduce

`nix-build -A pt2-clone.tests`

### Expected behavior

```
[...]
machine: must succeed: xwininfo -root -tree | sed 's/.*0x[0-9a-f]* \"\([^\"]*\)\".*/\1/; t; d'
(0.08 seconds)
(3.31 seconds)
machine: must succeed: sleep 5
(5.10 seconds)
machine: sending monitor command: screendump /build/tmp2_cesjz4/ppm
machine: making screenshot /nix/store/yscyb0ymgmn8nshkb260gmqca05krsg0-vm-test-run-pt2-clone/screen.png
machine: sending monitor command: screendump /nix/store/yscyb0ymgmn8nshkb260gmqca05krsg0-vm-test-run-pt2-clone/screen.png.ppm
(0.15 seconds)
(62.30 seconds)
test script finished in 62.45s
cleanup
kill machine (pid 8)
machine # qemu-system-x86_64: terminating on signal 15 from pid 6 (/nix/store/8dxxjbiyxwkvh53q5kh6nydla2anacgi-python3-3.9.6/bin/python3)
(0.09 seconds)
/nix/store/yscyb0ymgmn8nshkb260gmqca05krsg0-vm-test-run-pt2-clone
```

### Additional context

I ran a bisect and it looks like 1640359f333d8af40bf1f3e7961943ea04c6d1a1 (PR #142747) is the cause

```
1640359f333d8af40bf1f3e7961943ea04c6d1a1 is the first bad commit
commit 1640359f333d8af40bf1f3e7961943ea04c6d1a1
Author: Janne Heß <janne@hess.ooo>
Date:   Sun Oct 24 15:46:45 2021 +0200

    nixos/test-runner: Fix execute() flakiness
    
    Instead of using the magic string, we now just base64-encode everything
    and check for a newline.

 .../development/writing-nixos-tests.section.md     |  5 +++
 .../development/writing-nixos-tests.section.xml    |  8 ++++-
 nixos/lib/test-driver/test-driver.py               | 40 +++++++++++++++-------
 nixos/tests/hibernate.nix                          |  4 +--
 nixos/tests/kexec.nix                              |  2 +-
 5 files changed, 43 insertions(+), 16 deletions(-)
```

### Notify maintainers

Author of the above commit: @dasJ

### Metadata
Please run `nix-shell -p nix-info --run "nix-info -m"` and paste the result.

```console
[user@system:~]$ nix-shell -p nix-info --run "nix-info -m"
 - system: `"x86_64-linux"`
 - host os: `Linux 5.10.48, NixOS, 21.05.1412.cf59fbd5396 (Okapi)`
 - multi-user?: `yes`
 - sandbox: `yes`
 - version: `nix-env (Nix) 2.4pre20210601_5985b8b`
 - channels(root): `"nixos-21.05.1412.cf59fbd5396"`
 - channels(fgaz): `"nixpkgs-21.11pre326821.509d236edf0, stable-20.09.4407.1c1f5649bb9"`
 - nixpkgs: `/home/fgaz/.nix-defexpr/channels/nixpkgs`
```

Maintainer information:
```yaml
# a list of nixpkgs attributes affected by the problem
attribute:
- nixosTests.pt2-clone
- nixosTests.ft2-clone
- nixosTests.shattered-pixel-dungeon
- nixosTests.xterm
# a list of nixos modules affected by the problem
module:
```

### LINKED ISSUE #144694: nfs.simple and containers-imperative tests are broken
### Steps To Reproduce
1. `nix build -L -f nixos/tests/containers-imperative.nix`
or 
1. `nix build -L -f nixos/tests/nfs/simple.nix`

Tests will run forever

### Expected behavior
Tests run

### Additional context

Failure was introduced in 1640359f333d8af40bf1f3e7961943ea04c6d1a1 according to my bisection.
### Notify maintainers

@dasJ

### PR DESCRIPTION
**Pull request by roberth**
###### Motivation for this change

Fix #144613
Fix: #144694 

This doubles down on the new waiting for stdout behavior. Avoiding the new behavior seems unnecessarily tricky.

cc @dasJ @fgaz 

###### Things done

- Built on platform(s)
  - [ ] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/#sec-conf-file))
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review wip"`
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [21.11 Release Notes (or backporting 21.05 Release notes)](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#generating-2111-release-notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

### CODE REVIEW

**Comment by dasJ** [comment]
I really need some sleep so I'll take a look at this tomorrow. There's also `nixos.tests.nfs3.simple.x86_64-linux` which fails for probably the same reason

**Comment by lovesegfault** [comment]
I've tested this against all previously failing tests and it fixes all of them.

```
nix build --print-build-logs -f nixos/tests/pt2-clone.nix
nix build --print-build-logs -f nixos/tests/ft2-clone.nix
nix build --print-build-logs -f nixos/tests/shattered-pixel-dungeon.nix
nix build --print-build-logs -f nixos/tests/xterm.nix
nix build --print-build-logs -f nixos/tests/containers-imperative.nix
nix build --print-build-logs -f nixos/tests/nfs/simple.nix
```

**Comment by lovesegfault** [comment]
I'm going to merge this to unblock all the channels. It looks correct to me, I've tested it as much as I can, and I think any follow-ups can come in separate commits.

File: nixos/tests/libreswan.nix [hunk 1]
```diff
@@ -100,7 +100,7 @@ in
           Starts eavesdropping on Alice and Bob
           """
           match = "src host alice and dst host bob"
-          eve.execute(f"tcpdump -i br0 -c 1 -Avv {match} >/tmp/log &")
+          eve.execute(f"tcpdump -i br0 -c 1 -Avv {match} >/tmp/log >&2 &")
```
Comment (rnhmjoj) [inline nixos/tests/libreswan.nix:103]: This change is wrong and broke the test entirely.
I don't know why you didn't test this properly (was this done in a hurry to unblock the channel?), but you could have at least pinged the maitnainers.

Comment (roberth) [reply nixos/tests/libreswan.nix:103]: Yes, both the commits and the merge were made in a hurry. I hope you didn't lose much time on this. The rest of the PR must have saved many others time.

Comment (rnhmjoj) [reply nixos/tests/libreswan.nix:103]: I understand, thank you. It's just that it's hard to keep up with everything happening in nixpkgs.
I lost some time, but that mostly on me for not noticing this commit sooner.

**Pull request merged**